torch/csrc/jit/codegen/cuda/scheduler/pointwise.h

```diff
@@ -10,6 +10,130 @@ namespace jit {
+ * trivial case:
+ * T0[i0, b1]
+ * T1[b0, i1]
+ * T2[i0, i1] = T0 + T1
+ * If we scheduled T2 as 1-dimensional we would do something along the lines of
+ * merging i0 and i1 then splitting out a block and thread dimension. If i1 is
+ * greater than the thread dimension, then all threads would pull the same value
+ * from T0. However, they would all be pulling different values from T1. In this
+ * case we have perfect reuse of the broadcast dimension T0 but potentially no
+ * reuse of the broadcast dimension of T1. Potentially because if i1 isn't too
+ * big it should be efficiently cached in L2. If i1 is big, then by the time we
+ * increment the i0 dimension the i1 dimension will be pushed out of cache.
+ *
+ * Instead what we do is we map this to a two dimensional problem. Instead of
+ * having the schedule that merges the two dimensions, we'll actually leave the
+ * dimensions separate and we'll take i0, split it to BIDy, TIDy, and take i1
+ * and split it to BIDx and TIDx. Therefore we'll have a parallelization on T2
+ * like [BIDy, TIDy | BIDx, TIDx], where | denotes the separation of the
+ * original i0 and i1. This helps because all threads in the TIDx dimension will
+ * reuse the same value in the i0 dimension (holding BIDy and TIDy constant),
+ * all the threads in the TIDy dimension (holding BIDx, and TIDx constant) will
+ * reuse the same value in the i1 dimension. This reuse of values would prevent
+ * the number of times we'd have to pull unique values from T0 and T1. The same
+ * thing can be said for when incrementing BIDy, but since BIDy is strided on
+ * BIDx there's no effective increment of BIDy without incrementing BIDx. Since
+ * all threads are executed within a block we can effectively consider the block
+ * incrementing TIDx BDIMx times while holding TIDy constant and incrementing
+ * TIDy BDIMy times while holding TIDx constant. Since multiple BIDx's are
+ * running at the same time on the device we can consider a wave on the GPU of
+ * incrementing BIDx (wave number of times), while holding TIDy constant BDIMy *
+ * wave number of times.
+ *
+ * If instead we have a situation like:
+ * T0[i0, i1, b2]
+ * T1[i0, b1, i2]
+ * T2[i0, i1, i2] = T0 + T1
+ * It makes sense that the break point would be in position 2, between i1 and
+ * i2. This is because when we map [i0, i1 | i2] to [BIDy, TIDy| BIDx, TIDx]
+ * BIDx, and TIDx will access the same elements of T1, and TIDy will likely
+ * access the same elements of T0 (as long as i1 > BDIMy). Even if i1 on the
+ * order of BDIMy we'll only use two unique elemen*ts per increment of BIDx or
+ * TIDx. This means we'll still reuse many of the same values and limit the
+ * amount we need to reread values in T0 and T1.
+ *
+ * If instead we have:
+ * T0[i0, b1, i2]
+ * T1[b0, i1, i2]
+ * T2[i0, i1, i2] = T0 + T1
+ * The analysis gets a bit more complicated. First if i2 is very large and i0
+ * and i1 are relatively small it would make sense to have [i0, i1 | i2]. If b0
```

Comment:
Yeah, the thread bindings I think generally could use a review, should put a todo. It seems like we should probably prefer more square tiles, but should also figure out if there's inner, outer, or both inner and outer broadcasting when doing so. I think the heuristics could use a rework. We should probably look through the pytorch benchmarks to see if we can find some cases that would give us better information of which cases are important. I've kind of stamped 2D pointwise as being good enough for now because I don't know what cases are really important.